File: recyclerviewadapter/src/main/java/net/moltendorf/android/recyclerviewadapter/RecyclerViewAdapter.java
package net.moltendorf.android.recyclerviewadapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import rx.Observable;
import rx.subjects.PublishSubject;

import java.lang.reflect.ParameterizedType;
import java.util.*;

/**
 * More generic implementation of RecyclerView.
 * Can be used in various ways with minimal, if any, extension required.
 */
public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder> {
  private static final String TAG = "PlacesListAdapter";

  private Context context;
  private List dataSet;

  private Map<Class<?>, Integer> viewTypeLookup;
  private Factory[] viewHolderLookup;

  public RecyclerViewAdapter(Context context) {
    this(context, Collections.<Factory>emptySet());
  }

  public RecyclerViewAdapter(Context context, Set<Factory> factories) {
    this(context, factories, Collections.EMPTY_LIST);
  }

  public RecyclerViewAdapter(Context context, Set<Factory> factories, List dataSet) {
    Log.d(TAG, "PlacesListAdapter: Called.");

    this.context = context;
    this.dataSet = dataSet;

    setViewHolders(factories);
  }

  @Override
  public int getItemViewType(int position) {
    // Todo: add more helpful exception when no class to resource relationship exists.
    return viewTypeLookup.get(dataSet.get(position).getClass());
  }

  @Override
  public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    Factory factory = viewHolderLookup[viewType];
    ViewHolder viewHolder = factory.createViewHolder(context, parent);

    factory.subject.onNext(viewHolder);
    return viewHolder;
  }

  @Override
  public void onBindViewHolder(ViewHolder holder, int position) {
    holder.bindTo(dataSet.get(position), position);
  }

  @Override
  public int getItemCount() {
    return dataSet.size();
  }

  public void setViewHolders(Class<?>... holders) {
    Set<Factory> factories = new LinkedHashSet<>();

    outer:
    for (Class<?> holder : holders) {
      for (Class<?> possible : holder.getDeclaredClasses()) {
        if (Factory.class.isAssignableFrom(possible)) {
          try {
            factories.add((Factory) possible.newInstance());
          } catch (Exception exception) {
            throw new RuntimeException("Could not instantiate Factory.", exception);
          }

          continue outer;
        }
      }

      throw new RuntimeException("No ViewHolder Factory inside of " + holder.getSimpleName());
    }

    setViewHolders(factories);
  }

  public void setViewHolders(Set<Factory> factories) {
    // Create lookups.
    int size = factories.size();
    viewTypeLookup = new HashMap<>(size);
    viewHolderLookup = new Factory[size];

    int i = 0;
    for (Factory factory : factories) {
      viewTypeLookup.put(factory.getDataClass(), i);
      viewHolderLookup[i] = factory;

      ++i;
    }
  }

  public void changeDataSet(List dataSet) {
    this.dataSet = dataSet;

    notifyDataSetChanged();
  }

  abstract public static class ViewHolder<T> extends RecyclerView.ViewHolder {

    protected Context context;
    protected T object;
    protected int position;

    public ViewHolder(Context context, ViewGroup viewGroup, int resource) {
      super(LayoutInflater.from(context).inflate(resource, viewGroup, false));

      this.context = context;
    }

    private void bindTo(Object object, int position) {
      this.object = (T) object;
      this.position = position;

      bindTo();
    }

    abstract public void bindTo();
  }

  abstract public static class Factory<T extends ViewHolder> {
    private Class<?> dataClass;
    private PublishSubject<T> subject = PublishSubject.create();

    public Factory() {
      Class<?> viewHolderClass = (Class<?>) ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0];
      dataClass = (Class<?>) ((ParameterizedType) viewHolderClass.getGenericSuperclass()).getActualTypeArguments()[0];
    }

    public Class<?> getDataClass() {
      return dataClass;
    }

    public Observable<T> created() {
      return subject;
    }

    @Override
    public final boolean equals(Object o) {
      if (this == o) {
        return true;
      }

      if (o instanceof Factory) {
        Factory factory = (Factory) o;

        return dataClass.equals(factory.dataClass);
      }

      return false;
    }

    @Override
    public final int hashCode() {
      return dataClass.hashCode();
    }

    abstract public T createViewHolder(Context context, ViewGroup parent);
  }
}

File: recyclerviewadapter/src/main/kotlin/net/moltendorf/android/recyclerviewadapter/RecyclerViewAdapter.kt
package net.moltendorf.android.recyclerviewadapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import java.lang.reflect.ParameterizedType
import java.util.*

/**
 * More generic implementation of RecyclerView.
 * Can be used in various ways with minimal, if any, extension required.
 */
class RecyclerViewAdapter(private val context: Context) : RecyclerView.Adapter<RecyclerViewAdapter.ViewHolder<Any>>() {
  private var dataSet: List<Any> = Collections.emptyList();
  private var typeLookup: Map<Any, Int> = HashMap()
  private var factoryLookup = ArrayList<Factory<ViewHolder<Any>>>(0)

  override fun getItemViewType(position: Int): Int {
    // @todo Add more helpful exception when no class to resource relationship exists.
    val data = dataSet[position];

    val type = if (data is ViewType) {
      typeLookup[data.type] ?: typeLookup[data.javaClass]
    } else {
      typeLookup[data.javaClass]
    }

    return type ?: -1;
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<Any> {
    return factoryLookup[viewType].createViewHolder(context, parent) // @todo Note above @todo: This may throw due to that.
  }

  override fun onBindViewHolder(holder: ViewHolder<Any>, position: Int) {
    holder.bindTo(dataSet[position], position)
  }

  override fun getItemCount(): Int {
    return dataSet.size
  }

  fun setViewHolders(vararg holders: Class<*>) {
    val factories = LinkedHashSet<Factory<ViewHolder<Any>>>()

    outer@for (holder in holders) {
      for (possible in holder.declaredClasses) {
        if (Factory::class.java.isAssignableFrom(possible)) {
          try {
            factories.add(possible.newInstance() as Factory<ViewHolder<Any>>)
          } catch (exception: Exception) {
            throw RuntimeException("Could not instantiate Factory.", exception)
          }

          continue@outer
        }
      }

      throw RuntimeException("No ViewHolder Factory inside of " + holder.simpleName)
    }

    setViewHolders(factories)
  }

  fun setViewHolders(factories: Set<Factory<ViewHolder<Any>>>) {
    // Create lookups.
    val size = factories.size
    val typeLookup = HashMap<Any, Int>(size)
    val factoryLookup = ArrayList<Factory<ViewHolder<Any>>>(size)

    var i = 0
    for (factory in factories) {
      typeLookup[factory.type] = i
      factoryLookup[i] = factory

      ++i
    }

    this.typeLookup = typeLookup;
    this.factoryLookup = factoryLookup;
  }

  fun changeDataSet(dataSet: List<Any>) {
    this.dataSet = dataSet

    notifyDataSetChanged()
  }

  abstract class Factory<T : ViewHolder<Any>> {
    open val type: Any = (((javaClass.getGenericSuperclass() as ParameterizedType).actualTypeArguments[0] as Class<*>).genericSuperclass as ParameterizedType).actualTypeArguments[0] as Class<*>

    override fun equals(other: Any?): Boolean {
      if (this === other) {
        return true
      }

      if (other is Factory<*>) {
        return type == other.type
      }

      return false
    }

    override fun hashCode(): Int {
      return type.hashCode()
    }

    abstract fun createViewHolder(context: Context, parent: ViewGroup): T
  }

  abstract class ViewHolder<T>(protected var context: Context, viewGroup: ViewGroup, resource: Int) : RecyclerView.ViewHolder(LayoutInflater.from(context).inflate(resource, viewGroup, false)) {
    protected var data: T? = null
    protected var index: Int = 0

    internal fun bindTo(data: Any, index: Int) {
      this.data = data as? T
      this.index = index

      bindTo()
    }

    abstract fun bindTo()
  }

  companion object {
    private val TAG = "PlacesListAdapter"
  }
}
